### ISSUE TITLE
Unhandled rejection when using newrelic node agent for serverless monitoring and ioredis with a redis cluster

### ISSUE THREAD

**Issue by ininrobu** [newrelic/node-newrelic#566]
When using the newrelic node agent, specifically when instrumenting AWS Lambda functions, along with ioredis and a redis cluster I see an unhandled rejection in certain error scenarios.  The AWS Lambda runtime will exit in response to unhandled rejections which makes it impossible for me to handle the error scenarios and use fallback logic.

Note that this is specific to redis clusters - unclustered redis does not have this issue.  The broad class of error scenarios that cause this are when ioredis can't connect to the redis cluster.  In that scenario, there is a background `cluster slots` command that ioredis runs which eventually gets its associated promise rejected, and that rejection is what causes the unhandled rejection.  This is because that promise is forked by the newrelic agent when it intercepts it [here](https://github.com/newrelic/node-newrelic/blob/main/lib/shim/shim.js#L1915).  The forking is due to the fact that the promise ioredis stores with the command isn't the tip of the promise chain; it's one link up due to ioredis' use of `standard-as-callback`, specifically [this](https://github.com/luin/asCallback/blob/master/lib/index.ts#L41).

When the `cluster slots` promise is rejected, the fork created in `standard-as-callback` has a rejection handler registered so things work as expected on that fork, but the newrelic fork now has only a `finally` handler registered, causing the unhandled rejection event to be raised.

To give some real world context to why this is a problem, I ran into this because my Lambda function, which uses redis as a cache tier, couldn't reach redis for a brief period.  The results were Lambda execution errors and failed operations despite the fact that I had logic in place to handle cache layer failures and fall back to the home of record.  The fallback logic worked as expected, however the unhandled rejection caused by the newrelic instrumentation caused the node runtime to exit, resulting in the errors and failed operations.  Obviously having a function that errors out every time there's a cache tier blip isn't going to work.

I've attached a simple project to reproduce the issue, which can be run locally.  Unzip it, using node 12+ run `npm ci`, then `npm run works` to see how it works without instrumentation and `npm run boom` to see how it fails with instrumentation.

[newrelic-unhandled-rejections-repro.zip](https://github.com/newrelic/node-newrelic/files/5673968/newrelic-unhandled-rejections-repro.zip)

**Comment by michaelgoin** [comment]
Hi @ininrobu ,

Sorry for the troubles and thank you for the reproduction. We'll take a look and keep you posted.

**Comment by nijotz** [comment]
I ran out of time looking into this, but here's what I have done so far:

I started up a redis cluster using the latest release
```bash
tar xzvf redis-stable.tar.gz
make
cd utils/create-cluster/
./create-cluster start
./create-cluster create
```

This starts a cluster on 30001, so I modified the port in the repro.

Then I ran the repro:
```bash
$ npm run boom | grep NR_LAMBDA | jq '.[2]' | sed 's/"//g' | base64 -d | gunzip | jq '.data.analytic_event_data[2]'
[
  [
    {
      "webDuration": 0.021866972000000002,
      "timestamp": 1607651635762,
      "name": "OtherTransaction/Function/function-name",
      "duration": 0.021866972000000002,
      "totalTime": 0.023857294015625002,
      "type": "Transaction",
      "error": false,
      "databaseDuration": 0.026466763,
      "databaseCallCount": 6
    },
    {},
    {
      "aws.lambda.arn": "function-arn",
      "aws.requestId": "request-id",
      "aws.lambda.coldStart": true
    }
  ]
]
```

I'm not seeing an error immediately, but I could be looking in the wrong place, or I could be using different versions. I'll look at this again tomorrow.

**Comment by ininrobu** [comment]
@nijotz The key is to _not_ have redis running, to simulate the cache tier not being available for whatever reason.  In the case or AWS-hosted redis (Elasticache) that can happen due to hardware failure-induced failover or AWS outages, both of which my team has encountered.

**Comment by nijotz** [comment]
Oh right right. I wanted to be able to see it working properly and it breaking, and it took me awhile to figure out the redis cluster setup that I forgot the point was to turn it OFF too, haha. Thanks for the guidance, I'll look at it some more today.

**Comment by nijotz** [comment]
Hey @ininrobu,

I had a chance to look at this some more. It seems that I'm able to prevent the unhandled promise rejection by putting a `process.exit()` on line 49 in the `.then()` on the handler. I was looking through the ioredis source to see what was throwing the unhandled rejection, and it looks like an error that is [thrown on close](https://github.com/luin/ioredis/blob/515d9eaee8e2be0f31dc3fbf2264718bee2343f5/lib/cluster/index.ts#L320) and I couldn't catch it no matter how many error handlers I tried to setup on the redis object, the stream object, etc. It seems that there's some promises hanging around after teardown of the redis object perhaps? Not sure why this only happens with the newrelic module enabled. Maybe the agent being enabled results in a reference to those promises sticking around, and even after completion of the script node waits for the promises to finish and they error. Versus without the agent, the redis client is cleaned up properly. Node scripts with top-level promises have always acted funny for me...

Anyway, not sure what we can do on the newrelic side to detect this scenario, but at least there's a workaround. Let me know what you think and if this works for you.

Thanks,
Nick

**Comment by ininrobu** [comment]
Thanks @nijotz, unfortunately exiting the process isn't a usable workaround.  That's effectively what happens now - node Lambdas already have an unhandled rejection event handler registered that causes the function to exit with an error.  The trouble is that these redis-related failures shouldn't be fatal - my functions have fallback logic and should continue to operate even when redis is unavailable.  Exiting the process makes that fallback logic useless.

Regarding the actual error causing the unhandled rejection, it's not the one you linked (that one is handled and bubbles up to a `Failed to refresh slots cache` error which can be caught and handled), it's [this one that's used to flush the cache when the socket disconnects](https://github.com/luin/ioredis/blob/515d9eaee8e2be0f31dc3fbf2264718bee2343f5/lib/redis/event_handler.ts#L214).  As I mentioned in the initial description, there's a background `cluster slots` command queued up which gets rejected in response to that `flushQueue` call, causing the unhandled rejection.

As for why it only happens when the newrelic module is enabled, as I mentioned in the initial description it's because the newrelic module forks the ioredis command's promise chain, leaving a fork with no rejection handler registered.

I'm honestly not sure how the newrelic module could handle this without affecting other modules.  Adding a no-op rejection handler in addition to the finally handler would obviously fix this ioredis issue perfectly, but would potentially swallow unhandled rejections from other modules.

If there were an easy way to disable the ioredis instrumentation, that would be a viable workaround for me.

**Comment by nijotz** [comment]
Oh right right. I think I went down the rabbithole of our promise-handling code, and when I got back out I forgot the point was to *not* exit, not exit cleanly. Yes, it is a difficult task trying to determine if the promise being handled has been forked and if the other fork has or will be handled cleanly. I'll continue to look into that and if I find anything I will report back here.

But as for disabling instrumentation for certain modules, we have an experimental branch [here](https://github.com/newrelic/node-newrelic/tree/experimental/disable-instrumentation#experimental-configuration-disabling) that allows that. I'll also check with the team and see if there's other ways to disable instrumentation on specific modules.

Thanks @ininrobu!
